### ISSUE TITLE
WPGraphQL v1.0

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql#1357]
# WPGraphQL is turning 1.0!!! 🥳  🎉  🍰  🎈 🍾  

The intent of this issue is to outline what the path to 1.0 looks like and how things will change (or be the same) for WPGraphQL after 1.0. 

## The Journey to 1.0

WPGraphQL was started in November 2016 (with [no commit message](https://github.com/wp-graphql/wp-graphql/commit/6723f8cdf16b85cc827fe7ec1cbcb10aa02fa616) 🤣 ) and has come a long way in that time. 

By the end of 2017 the codebase was pretty robust and provided a Schema to interact with many parts of WordPress. Posts, Pages, Custom Post Types, Tags, Categories, Custom Taxonomies, Comments and Nav Menus. 

[Qz.com](https://qz.com) was the first site to go to production with WPGraphQL, even before _I_ was using it in production! A few years later and they are still running on WPGraphQL! You can learn more about the early days of their stack from [this presentation](https://www.youtube.com/watch?v=JvkDdYAA64A) from December 2017.

In early 2018, my team took WPGraphQL into production on a network of 54 large WordPress sites. WPGraphQL was used as a syndication engine for our network of newspapers. Unlike most GraphQL users, we didn't initially have a JavaScript or native mobile UI that was using GraphQL, we had PHP WordPress servers syndicating content and GraphQL solved a lot of pain points we had been facing when using REST. 

By the end of 2018, more than 100 sites were using WPGraphQL in some way. 

Now, WPGraphQL seems to be running all over the place!

This is a non-comprehensive list of sites using WPGraphQL in production in some way: 

- [DenverPost.com](https://denverpost.com)
- [TwinCities.com](https://twincities.com)
- [MercuryNews.com](https://mercurynews.com)
- [SiliconValley.com](https://siliconvalley.com)
- [QZ.com](https://qz.com)
- [MeowWolf.com](https://meowwolf.com)
- [ThePlayersTribune.com](https://theplayerstribune.com)
- [ApolloGraphQL.com](https://apollographql.com/blog)
- [blog.creditkarma.com](https://blog.creditkarma.com)

According to [Packagist.org](https://packagist.org/packages/wp-graphql/wp-graphql) as of June 30, 2020 there are nearly 50,000 installs of WPGraphQL in the wild! 

In 2019 Credit Karma paid for a [security audit](https://www.pentestpartners.com/security-blog/pwning-wordpress-graphql/) on the plugin and all reported issues were resolved.

It's safe to say, WPGraphQL is production ready!

## So, why the wait for 1.0?

So, if WPGraphQL has been production ready for such a long time, why is is still not 1.0?

### Breaking Changes & WordPress 

The WordPress ecosystem has a strong commitment to backward compatibility. This means that once functionality is introduced to a WordPress plugin, it often sticks around _forever_. 

With that in mind, before tagging WPGraphQL 1.0, I wanted so badly to be at a place with the codebase & Schema that I couldn't predict anymore breaking changes. I wanted the GraphQL Schema to be "perfect" before tagging it 1.0, because I knew that whatever would be in the WPGraphQL Schema at 1.0 would need to remain in the Schema _forever_. The WordPress community is used to things working the same way "forever" and I wanted to play nice. 

### Breaking Changes & WPGraphQL

The reality is that for software to get better, sometimes it needs to make breaking changes. WPGraphQL will never be perfect, and it will _need_ to occasionally break to make things better. For example, when the GraphQL Specification updates to include Input Unions or Interfaces that can implement other Interfaces, WPGraphQL will need to break to keep up with the GraphQL Specification. There would be no way to keep GraphQL as we know it today and take advantage of tomorrow's GraphQL Spec without breaking. 

**_WPGraphQL turning 1.0 isn't a statement that there will never be breaking changes, instead it's a statement of stability and long term support._** 

So, how will breaking changes be handled in a post-1.0 world?

### Semantic Versioning
WPGraphQL has been following [Semver](https://semver.org/) practices for a few years. We will continue to follow Semver and let version numbers communicate meaning. The summary of Semver versioning is as follows: 

- **MAJOR** version when you make incompatible API changes,
- **MINOR** version when you add functionality in a backwards compatible manner, and
- **PATCH** version when you make backwards compatible bug fixes.

You can read more about the details of Semver at [semver.org](https://semver.org/)

### Seamless Upgrade Paths
When WPGraphQL needs to make breaking changes to the Schema, there are sometimes ways to make the changes and allow for seamless upgrade paths. 

What that means, is that WPGraphQL can introduce a new feature while keeping the old feature. For example, let's say the following (hypothetical) query were in the Schema and the `author` field returned the name of the Author: 

```
{
  book {
    author
  }
}
```

Let's say we realized that the Author should actually be a one-to-one connection to an object with edge space for relational context. We would want to change the schema to allow for the following query instead. 

```
{
  book {
     author {
         node {
             name
          }
     }
  }
}
```

If WPGraphQL made this change, any consumer application would break. 

So, what we can do is introduce the new feature like so: 

```
{
   book {
     author
     authoredBy {
        node {
            name
        }
     }
   }
}
```

We could introduce the new feature as a new field name `authoredBy` which returns the data in the shape we want. 

The end goal, however, is to have the `author` field return the `author.node.name` shape. 

So, with both fields (author and authoredBy) co-existing on the server, consumer applications could make use of GraphQL Aliases and update their code from: 

```
{
  book {
     author
  }
}
```

to: 

```
{
   book {
      author: authoredBy { # Consumer uses an alias to prepare for the upcoming change to the author field
          node {
             name
          }
      }
   }
}
```

Once consumers have updated their code, the server could then update once more to flip the old field name to the new shape. So, the next release would have the following possible: 

```
{
   book {
      author {
         node {
             name
          }
      }
      authoredBy {
        node {
            name
        }
     }
   }
}
```

So now, the old field `author` no longer returns the name directly, it now returns the new nested node/name shape. The client is using an alias to query the `authoredBy` field, so now the consumer should update their code from: 

```
{
   book {
      author: authoredBy { # Consumer uses an alias to prepare for the upcoming change to the author field
          node {
             name
          }
      }
   }
}
```

to: 

```
{
   book {
      author
          node {
             name
          }
      }
   }
}
```

Now, the consumer application is using the new shape of data and the old author field, and one more release would be able to remove the temporary `authoredBy` field that was used for seamless upgrades.

## Timeline and What's Left

There's a [short-list](https://github.com/wp-graphql/wp-graphql/milestone/24) of predictable breaking changes that remain to be addressed for the 1.0 release candidate. 

The plan is to have a Pull Request open with these changes in the next few days. 

Then, there are some bugs, features and enhancements that have been discussed as [candidates](https://github.com/wp-graphql/wp-graphql/milestone/4) for the final 1.0 release. 

These will be worked on as 1.0-rc.* releases as we work toward the final v1.0. 

Once the issues in the 1.0 milestone are addressed, 1.0 will be released! 

## The path to WPGraphQL 2.0
Prior to v1.0 we've not supported previous releases. The expectation has been that users are using the latest release, and new features/bugfixes will _only_ be applied to the latest release.

Once WPGraphQL v1.0 is released, any new features, bugfixes and enhancements will be supported in 1.* releases. 

A 2.0 branch will be opened and any breaking change features will be added to that. This will allow folks that want to be stable to continue using 1.* for a while and folks that want to be on the bleeding edge and start testing breaking changes early can start using and testing 2.0 before it's officially released. 

## How can you help?

If you want to get involved with WPGraphQL, here are some good ways: 

- Join the WPGraphQL Community in [Slack](https://wpgql-slack.herokuapp.com/)
- Review open issues in Github, specifically those in the 1.0 milestone. Interested in helping resolve one? Comment on the issue or reach out in [Slack]((https://wpgql-slack.herokuapp.com/))
- Follow [WPGraphQL](https://twitter.com/wpgraphql) on Twitter
- Contribute to the [WPGraphQL Docs](https://github.com/wp-graphql/wp-graphql/tree/develop/docs)
- Help inspire others by writing tutorials, case studies, etc about how you're using WPGraphQL – and let us know so we can share!

**Comment by whiteboypls** [comment]
Thank you so much for your contribution to the community!

**Comment by jasonbahl** [comment]
Closed by #1585
